rustfmt-core/rustfmt-lib/src/formatting/report.rs

```diff
@@ -0,0 +1,188 @@
+use crate::result::{ErrorKind, FormatError};
+use crate::FileName;
+use crate::NewlineStyle;
+
+/// Reports on any issues that occurred during a run of Rustfmt.
+///
+/// Can be reported to the user using the `Display` impl on [`FormatReportFormatter`].
+#[derive(Debug, Clone)]
+pub struct FormatReport {
+    format_result: Rc<RefCell<BTreeMap<FileName, FormatResult>>>,
+    ignored_files: Rc<RefCell<BTreeSet<FileName>>>,
+}
+
+/// The result of formatting, including the formatted text and various
+/// errors and warning arose while formatting.
+#[derive(Debug, Clone, Default)]
+pub struct FormatResult {
+    original_snippet: Option<String>,
+    formatted_snippet: FormattedSnippet,
+    format_errors: HashSet<FormatError>,
+    newline_style: NewlineStyle,
+}
+
+/// The inclusive range of the input which was not formatted, represented by a pair of line numbers.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct NonFormattedRange {
+    lo: usize,
+    hi: usize,
+}
+
+impl FormatResult {
+    pub(crate) fn success(
+        snippet: String,
+        non_formatted_ranges: Vec<NonFormattedRange>,
+        original_snippet: Option<String>,
+        newline_style: NewlineStyle,
+    ) -> Self {
+        let formatted_snippet = FormattedSnippet {
+            snippet,
+            non_formatted_ranges,
+        };
+        FormatResult {
+            original_snippet,
+            formatted_snippet,
+            format_errors: HashSet::new(),
+            newline_style,
+        }
+    }
+
+    pub(crate) fn all_errors(&self) -> impl Iterator<Item = &FormatError> {
+        self.format_errors.iter()
+    }
+
+    pub(crate) fn errors_excluding_macro(&self) -> impl Iterator<Item = &FormatError> {
+        self.format_errors
+            .iter()
+            .filter(|e| e.kind() != ErrorKind::MacroFormatError)
+    }
+
+    /// Return the newline style used to format the result.
+    pub fn newline_style(&self) -> NewlineStyle {
+        self.newline_style.clone()
+    }
+
+    pub fn original_text(&self) -> Option<&str> {
+        self.original_snippet.as_ref().map(|s| s.as_str())
+    }
+
+    pub fn formatted_text(&self) -> &str {
+        &self.formatted_snippet.snippet
+    }
+
+    pub(crate) fn formatted_snippet(&self) -> &FormattedSnippet {
+        &self.formatted_snippet
+    }
+}
+
+impl FormatReport {
+    pub fn new() -> FormatReport {
+        FormatReport {
+            format_result: Rc::new(RefCell::new(BTreeMap::new())),
+            ignored_files: Rc::new(RefCell::new(BTreeSet::new())),
+        }
+    }
+
+    /// Returns the result of formatting the given input, including the formatted text and
+    /// various warnings and errors encountered during formatting.
+    //
+    // NOTE: Avoid using this method internally: use `format_result_as_rc` instead.
+    pub fn format_result(&self) -> impl Iterator<Item = (&FileName, &FormatResult)> {
+        Ref::leak(RefCell::borrow(&self.format_result)).iter()
```

Comment:
Will this pose any potential issues for rustfmt on the stable channel if it lands before `cell::{Ref, RefMut}::leak` is stabilized upstream, or will the `cell_leak` feature gate suffice?

Reply:
I honestly don't know exactly, but I remember once using a nightly only feature was an issue with a beta channel build.

The reason for using `RefCell::leak` is to remove Rc and RefCell from the public interface as they will be removed in the part of refactoring. That being said, the refactoring won't happen anytime soon, so I will take a safety precaution and remove the usage of `RefCell::leak`.

Reply:
Yeah RefCell::leak seems like it would be very useful here, hopefully it will get stabilized soon!